examples/pytorch/NGCF/NGCF/main.py

```diff
@@ -0,0 +1,109 @@
+import torch
+import torch.optim as optim
+from model import NGCF
+from utility.batch_test import *
+from utility.helper import early_stopping
+from time import time
+import os
+
+def main(args):
+    # Step 1: Prepare graph data and device ================================================================= #
+    if args.gpu >= 0 and torch.cuda.is_available():
+        device = 'cuda:{}'.format(args.gpu)
+    else:
+        device = 'cpu'
+
+    g=data_generator.g
+    g=g.to(device)
+
+    # Step 2: Create model and training components=========================================================== #
+    args.mess_dropout = eval(args.mess_dropout)
+    args.layer_size = eval(args.layer_size)
+    lmbd = eval(args.regs)[0]
+    model = NGCF(g, args.embed_size, args.layer_size, args.mess_dropout, lmbd).to(device)
+    optimizer = optim.Adam(model.parameters(), lr=args.lr)
+
+    # Step 3: training epoches ============================================================================== #
+    t0 = time()
+    cur_best_pre_0, stopping_step = 0, 0
+    loss_loger, pre_loger, rec_loger, ndcg_loger, hit_loger = [], [], [], [], []
+    for epoch in range(args.epoch):
+        t1 = time()
+        loss, mf_loss, emb_loss = 0., 0., 0.
+        n_batch = data_generator.n_train // args.batch_size + 1
+
+        for idx in range(n_batch):
+            users, pos_items, neg_items = data_generator.sample()
+            u_g_embeddings, pos_i_g_embeddings, neg_i_g_embeddings = model(g, 'user', 'item', users,
+                                                                           pos_items,
+                                                                           neg_items)
+
+            batch_loss, batch_mf_loss, batch_emb_loss = model.create_bpr_loss(u_g_embeddings,
+                                                                              pos_i_g_embeddings,
+                                                                              neg_i_g_embeddings)
+            optimizer.zero_grad()
+            batch_loss.backward()
+            optimizer.step()
+
+            loss += batch_loss
+            mf_loss += batch_mf_loss
+            emb_loss += batch_emb_loss
+            
+
+        if (epoch + 1) % 10 != 0:
+            if args.verbose > 0 and epoch % args.verbose == 0:
+                perf_str = 'Epoch %d [%.1fs]: train==[%.5f=%.5f + %.5f]' % (
+                    epoch, time() - t1, loss, mf_loss, emb_loss)
+                print(perf_str)
+            continue
```

Comment:
Break the current epoch to miss the afterwards evaluation, which just need to be run every ten epochs. When "continue" is performed, it will end the current epoch and move to the next epoch.